lib/win32/certstore/certstore.rb

```diff
@@ -15,15 +15,46 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-require "win32/certstore/version"
-require 'win32/api/reserved_names'
+require 'store/crypto'
+require 'store/assertions'
+require_relative 'store_base'
+require 'pry'
 
 module Win32
-  module Certstore
-    include Win32::API::ReservedNames
+  class Certstore
+    include Win32::Store::Crypto
+    extend Win32::Store::Assertions
     include Chef::Mixin::WideString
+    extend Win32::Certstore::StoreBase
 
-    def open store_name
+    attr_accessor :store_name
+
+    def self.open(store_name)
+      validate_store(store_name)
+      @certstore_handle = self.new.send(:open, store_name)
+      if block_given?
+        yield self
+      else
+        self
+      end
+    end
+
+    def self.list
+      list = cert_list(@certstore_handle)
+      self.new.send(:close)
+      return list
+    end
+
+    def self.add(cert_file_path)
+      # validate_certificate(cert_file_path)
+      add = cert_add(@certstore_handle, cert_file_path)
+      self.new.send(:close)

```

Comment:
I don't understand this behavior. Why are we mixing instance methods and class methods? I'd expect something like

```
my_cert_store = Win32::CertStore.new
my_cert_store.open("Personal") # not sure this is a valid string?
my_cert_store.add(....)
```

Reply:
@btm - Above as per @mwrock comments, the class method of open could instantiate an instance of a store and return it. One would call list and other instance methods of that. open could also take a block so you could have:
```
Win32::Certstore.open("MY") do |store| 
    store.list
end
```
So here we are not creating any instance of Certstore and call an open method with that.(https://github.com/chef/win32-certstore/pull/4#discussion_r135021135)

Similarly while adding certificate we are not creating an instance of the Certstore class. We can add a certificate in certificate store as:
```
Win32::Certstore.open("MY") do |store| 
    store.add(certificate_file_path)
end
```
@btm, Please let me know if you want to achieve with an instance of Certstore for adding certificate in certificate store.

Reply:
I do think add and list should be instance methods as well as any other operation on a certstore. Once you obtain an instanece with open, you should operate on that instance.